core/primitives/src/shard_layout.rs

```diff
@@ -212,6 +202,114 @@ pub struct ShardLayoutV2 {
     version: ShardVersion,
 }
 
+/// Counterpart to `ShardLayoutV2` composed of maps with string keys to aid
+/// serde serialization.
+#[derive(serde::Serialize, serde::Deserialize)]
+struct SerdeShardLayoutV2 {
+    boundary_accounts: Vec<AccountId>,
+    shard_ids: Vec<ShardId>,
+    id_to_index_map: BTreeMap<String, ShardIndex>,
+    index_to_id_map: BTreeMap<String, ShardId>,
+    shards_split_map: Option<BTreeMap<String, Vec<ShardId>>>,
+    shards_parent_map: Option<BTreeMap<String, ShardId>>,
+    version: ShardVersion,
+}
+
+impl From<&ShardLayoutV2> for SerdeShardLayoutV2 {
+    fn from(layout: &ShardLayoutV2) -> Self {
+        let id_to_index_map =
+            layout.id_to_index_map.iter().map(|(k, v)| (k.to_string(), *v)).collect();
+
+        let index_to_id_map =
+            layout.index_to_id_map.iter().map(|(k, v)| (k.to_string(), *v)).collect();
+
+        let shards_split_map = layout
+            .shards_split_map
+            .as_ref()
+            .map(|map| map.iter().map(|(k, v)| (k.to_string(), v.clone())).collect());
+
+        let shards_parent_map = layout
+            .shards_parent_map
+            .as_ref()
+            .map(|map| map.iter().map(|(k, v)| (k.to_string(), *v)).collect());
+
+        Self {
+            boundary_accounts: layout.boundary_accounts.clone(),
+            shard_ids: layout.shard_ids.clone(),
+            id_to_index_map,
+            index_to_id_map,
+            shards_split_map,
+            shards_parent_map,
+            version: layout.version,
+        }
+    }
+}
+
+impl TryFrom<SerdeShardLayoutV2> for ShardLayoutV2 {
+    type Error = Box<dyn std::error::Error + Send + Sync>;
+
+    fn try_from(layout: SerdeShardLayoutV2) -> Result<Self, Self::Error> {
```

Comment:
mini nit: May unpack the layout as first step and then use the unpacked values directly? It may be a bit prettier and it would be more obvious that there isn't unnecessary cloning.